packages/@aws-cdk/aws-appmesh/lib/gateway-route-spec.ts

```diff
@@ -1,32 +1,222 @@
+   *
+   * @default - no match on method
    */
-  readonly prefixPath: string;
+  readonly method?: HttpRouteMatchMethod;
+
+  /**
+   * The query parameter to match on.
+   * All specified query parameter must match for the route to match.
+   *
+   * @default - no match on query parameter
+   */
+  readonly queryParameters?: QueryParameterMatch[];
 }
 
 /**
  * The criterion for determining a request match for this GatewayRoute
  */
 export interface GrpcGatewayRouteMatch {
   /**
-   * The fully qualified domain name for the service to match from the request
+   * Create service name based gRPC gateway route match.
+   *
+   * @default - no matching on service name.
+   */
+  readonly serviceName?: string;
+
+  /**
+   * Create host name based gRPC gateway route match.
+   *
+   * @default - No matching on host name.
+   */
+  readonly hostname?: GatewayRouteHostname;
+
+  /**
+   * Create metadata based gRPC gateway route match.
+   *
+   * @default - not matching on metadata.
+   */
+  readonly metadata?: HeaderMatch[];
+}
+
+/**
+ * Path and prefix properties for gRPC gateway route rewrite.
+ */
+export interface HttpGatewayRouteRewriteConfig {
+  /**
+   * GatewayRoute CFN configuration for gRPC gateway route path rewrite.
+   *
+   * @default - rewrite path to '/'.
+   */
+  readonly path?: CfnGatewayRoute.HttpGatewayRoutePathRewriteProperty;
+
+  /**
+   * GatewayRoute CFN configuration for gRPC gateway route prefix rewrite.
+   *
+   * @default - rewrite prefix to '/'.
+   */
+  readonly prefix?: CfnGatewayRoute.HttpGatewayRoutePrefixRewriteProperty;
+}
+
+/**
+ * Defines gRPC gateway route rewrite that can be defined in addition to host name.
+ */
+export abstract class HttpGatewayRouteRewrite {
+  /**
+   * The path to rewrite.
+   *
+   * @param exact The exact path to rewrite.
+   */
+  public static exactPath(exact: string): HttpGatewayRouteRewrite {
+    return new HttpGatewayRoutePathRewriteImpl(exact);
+  }
+
+  /**
+   * The default prefix used to replace the incoming route prefix when rewritten.
+   * When enabled, rewrites the matched path in Gateway Route to '/'.
+   * When disabled, does not rewrite to '/'.
+   *
+   * @param isDefault Boolean to select to enable or disable the default prefix.
+   *  Default value is enabled.
+   */
+  public static defaultPrefix(isDefault?: boolean): HttpGatewayRouteRewrite {
+    return new HttpGatewayRoutePrefixRewriteImpl({ defaultPrefix: isDefault == false ? 'DISABLED' : 'ENABLED' });
+  }
+
+  /**
+   * Replace the incoming route prefix when rewritten.
+   *
+   * @param value The value used to replace the incoming route prefix when rewritten.
+   */
+  public static customPrefix(value: string): HttpGatewayRouteRewrite {
+    return new HttpGatewayRoutePrefixRewriteImpl({ value: value } );
+  }
+
+  /**
+   * Return HTTP gateway route rewrite configuration.
    */
-  readonly serviceName: string;
+  abstract bind(scope: Construct): HttpGatewayRouteRewriteConfig;
+}
```

Comment:
**Pending to respond.**

Reply:
@skinny85 - I am not exactly sure what your question is. Are you asking why `HttpGatewayRoutePathRewrite` class is created?

If that is the case, `Path` and `Prefix` properties in the [`HttpGatewayRouteRewrite`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-gatewayroute-httpgatewayrouterewrite.html) object are union-type properties and can be selected just one.

Your `path` concept from `match` also applies here so `Path` and `Prefix` relationship can be written something like:
```
HttpGatewayRoutePathRewrite: {
  Exact : String,  // This represents exact path 
  DefaultPrefix : String,  // This is ENABLED or DISABLED. ENABLED means rewrite to prefix with '/'
  Value : String,  // This is custom prefix value.
},
```

Did I answer your question?

Reply:
Can you show me some example values of paths, and how different combinations of `match` and `rewrite` change that path, and what is the final result?

Reply:
This is the high-level process for handing the in-coming request to Virtual Gateway:
1. Request reaches a Virtual Gateway
2. **Gateway routes** associated to the Virtual Gateway specify the path to be matched on based on `match`
3. If matched, request is auto-rewritten based on `rewrite`.

For example:
// GIVEN
- In-coming request: `my-service.local/metrics`
- `match`: `HttpRoutePathMatch.prefix(`/metrics`)
// WHEN
- `rewrite`: HttpGatewayRoutePathRewrite.defaultPrefix(true)
//THEN
- End-result: in-coming request is rewritten to `/`.

//WHEN 
- `rewrite`: HttpGatewayRoutePathRewrite.defaultPrefix(false)
//THEN
- end-result: in-coming request is retained to '/metrics'.

Reply:
Thanks for the explanation! This is super helpful.

Can you also elaborate on the result for the following cases?

```
// GIVEN
- In-coming request: `my-service.local/metrics`
- `match`: `HttpRoutePathMatch.prefix(`/metrics`)
```

What is the result for:

1. `rewrite: HttpGatewayRoutePathRewrite.exact('/mypath')`
2. `rewrite: HttpGatewayRoutePathRewrite. customPrefix('/mypath')`

?

Reply:
Continuing the thread in [this comment](https://github.com/aws/aws-cdk/pull/15395#discussion_r665601320).

Reply:
```
// GIVEN
In-coming request: `my-service.local/metrics`
`match`: `HttpRoutePathMatch.exact(`/metrics`)

// WHEN
rewrite: HttpGatewayRoutePathRewrite.exact('/mypath')

// THEN
end-result: `my-service.local/mypath`
```

```
// GIVEN
In-coming request: `my-service.local/metrics_service`
`match`: `HttpRoutePathMatch.prefix(`/metrics`)

// WHEN
rewrite: HttpGatewayRoutePathRewrite.customPrefix('/mypath')

// THEN
end-result: In-coming request: `my-service.local/mypath_service`
```

Reply:
Thanks!

Also confirmed offline that the following case:


```
// GIVEN
In-coming request: `my-service.local/metrics`
`match`: `HttpRoutePathMatch.exact(`/metrics`)
// WHEN
rewrite: HttpGatewayRoutePathRewrite.customPrefix('/mypath')
```

Is not allowed.

Given that, we decided to get rid of the `HttpGatewayRoutePathRewrite` class completely, and simply allow providing the re-write path as the second argument to static factory methods in `HttpRoutePathMatch`.

Reply:
@skinny85 - I actually retained `HttpGatewayRoutePathRewrite` for only `prefix` rewrite since `prefix` rewrite allows to define `defaultPrefix` and `customPrefix`.

Reply:
Can you explain how is `defaultPrefix` different than just a `/` `customPrefix` in that case?